packages/opencensus-instrumentation-http2/src/http2.ts

```diff
@@ -0,0 +1,283 @@
+/**
+ * Copyright 2018, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {HttpPlugin} from '@opencensus/instrumentation-http';
+import {types} from '@opencensus/opencensus-core';
+import {classes} from '@opencensus/opencensus-core';
+import {logger} from '@opencensus/opencensus-core';
+import * as http from 'http';
+import * as http2 from 'http2';
+import * as net from 'net';
+import * as shimmer from 'shimmer';
+import * as tls from 'tls';
+import * as url from 'url';
+import * as uuid from 'uuid';
+
+export type Http2Module = typeof http2;
+export type ConnectFunction = typeof http2.connect;
+type RequestFunction =
+    (this: http2.ClientHttp2Session, headers?: http2.OutgoingHttpHeaders,
+     options?: http2.ClientSessionRequestOptions) => http2.ClientHttp2Stream;
+export type CreateServerFunction = typeof http2.createServer;
+
+/** Http2 instrumentation plugin for Opencensus */
+export class Http2Plugin extends HttpPlugin {
+  server: http2.Http2Server|http2.Http2SecureServer;
+  client: http2.ClientHttp2Session;
```

Comment:
If `server` and `client` are singular instances of HTTP2 servers and clients... how will you handle multiple instances?

Reply:
`server` and `client` always point toward the most-recently constructed instances, so won't `unpatch` only unpatch those, and not previously constructed ones?

A secondary concern here is that the `clientUrl` value will not be accurate if there are concurrent client requests. It seems like it will always hold the URL of the most recently connected client, but its value is used when a client request ends.

Generally, I think all of these should not be member variables. They aren't properties of the plugin, after all.

Reply:
In fact `clientUrl` where only holding the url from the first client instance. I removed the `clientUrl` and the unpatch of `Connect`, now, every time a new `client` is created `request` is patched passing the right one.

I don't think `server` and `client` should be properties of plugin too. I only used them to later unpatch `request` and `emit`. I removed them, but now only the `client` and `server` constructors will be unwrapped, any existing `client` or `server` instances will still trace.